### PR TITLE
ENH: support custom AES endpoint with SSL certificate file

### PR DESCRIPTION
**Pull request by chenqi0805**
*Issue #, if available:*
#341 

*Description of changes:*
This PR tackles option 1 in #341 . As discussed offline with @kowshikn , for now we do not need ARN support at the data-prepper level. The data-prepper built from this PR has been run on the VPC EC2 instance that connects with AES configured with custom endpoint.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: data-prepper-plugins/elasticsearch/README.md [hunk 1]
```diff
@@ -42,13 +42,36 @@ pipeline:
 
 The elasticsearch sink will reserve `otel-v1-apm-service-map` as index for record ingestion.
 
+### Amazon Elasticsearch Service (AES)
```
Comment (wrijeff) [inline data-prepper-plugins/elasticsearch/README.md:45]: nit - would replace `AES` with `Amazon ES` throughout

File: data-prepper-plugins/elasticsearch/src/test/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfigurationTests.java [hunk 2]
```diff
@@ -118,6 +118,26 @@ public void testCreateClientWithAWSSigV4DefaultRegion() throws IOException {
...
     }
 
+    @Test
+    public void testCreateClientWithAWSSigV4AndInsecure() throws IOException {
+        final PluginSetting pluginSetting = generatePluginSetting(
+                TEST_HOSTS, null, null, null, null, true, null, null, true);
+        final ConnectionConfiguration connectionConfiguration =
+                ConnectionConfiguration.readConnectionConfiguration(pluginSetting);
+        assertEquals("us-east-1", connectionConfiguration.getAwsRegion());
+        assertTrue(connectionConfiguration.isAwsSigv4());;
```
Comment (wrijeff) [inline data-prepper-plugins/elasticsearch/src/test/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfigurationTests.java:128]: Are there any other assertions we can make on these 2 scenarios? Currently they make the same assertions even though the code paths are different. If not then integ tests may be the way to go.

Comment (chenqi0805) [reply data-prepper-plugins/elasticsearch/src/test/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfigurationTests.java:128]: `insecure` flag is not exposed. integ test might be the only way to manifest.

File: data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfiguration.java [hunk 3]
```diff
@@ -180,8 +180,14 @@ private void attachAWSSigV4Callback(final RestClientBuilder restClientBuilder) {
...
-    restClientBuilder.setHttpClientConfigCallback(httpClientBuilder -> httpClientBuilder.addInterceptorLast(
-            httpRequestInterceptor));
+    final SSLContext sslContext = certPath != null ? getCAStrategy(certPath) : getTrustAllStrategy();
+    restClientBuilder.setHttpClientConfigCallback(httpClientBuilder -> {
+      httpClientBuilder.setSSLContext(sslContext).addInterceptorLast(httpRequestInterceptor);
+      if (this.insecure) {
+        httpClientBuilder.setSSLHostnameVerifier(NoopHostnameVerifier.INSTANCE);
+      }
+      return httpClientBuilder;
+    });
```
Comment (kowshikn) [inline data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfiguration.java:190]: This code is repeated in both AWSSigV4 and username methods, can you not attach. the ssl in the `createClient` instead?

**Review by kowshikn — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfiguration.java (hunk 3 above)
Comment (chenqi0805) [reply data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfiguration.java:190]: Due to the callback structure, I refactored out `attachSSLContext(final HttpAsyncClientBuilder httpClientBuilder)` and invoked it in both.

File: data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfiguration.java [hunk 4]
```diff
@@ -170,7 +172,7 @@ public RestHighLevelClient createClient() {
     return new RestHighLevelClient(restClientBuilder);
   }
 
-  private void attachAWSSigV4Callback(final RestClientBuilder restClientBuilder) {
+  private void attachAWSSigV4CallbackAndSSLContext(final RestClientBuilder restClientBuilder) {
```
Comment (kowshikn) [inline data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfiguration.java:175]: nit: `attachSigV4`

File: data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfiguration.java [hunk 5]
```diff
@@ -180,8 +182,11 @@ private void attachAWSSigV4Callback(final RestClientBuilder restClientBuilder) {
...
-    restClientBuilder.setHttpClientConfigCallback(httpClientBuilder -> httpClientBuilder.addInterceptorLast(
-            httpRequestInterceptor));
+    restClientBuilder.setHttpClientConfigCallback(httpClientBuilder -> {
+      httpClientBuilder.addInterceptorLast(httpRequestInterceptor);
+      attachSSLContext(httpClientBuilder);
+      return httpClientBuilder;
+    });
   }
 
   private void attachSSLUsernameContext(final RestClientBuilder restClientBuilder) {
```
Comment (kowshikn) [inline data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/ConnectionConfiguration.java:192]: nit: `attachUserCredentials`

**Review by kowshikn — APPROVED** [review]

File: data-prepper-plugins/elasticsearch/README.md (hunk 1 above)
Comment (yadavcbala) [reply data-prepper-plugins/elasticsearch/README.md]: Apparently we have legal restrictions on the name - please use Amazon Elasticsearch Service; not AES or Amazon ES even

Comment (chenqi0805) [reply data-prepper-plugins/elasticsearch/README.md]: @yadavcbala Thanks for call out!

**Review by yadavcbala — APPROVED** [review]

File: data-prepper-plugins/elasticsearch/README.md (hunk 1 above)
Comment (wrijeff) [reply data-prepper-plugins/elasticsearch/README.md]: @yadavcbala: Oh interesting, do you know where that's coming from? We recently published the blog post with legal's approval, the guidance given was to use Amazon Elasticsearch Service first, then optionally Amazon ES through the rest of the document. We can sync on this offline

**Review by wrijeff — APPROVED** [review]
👍

File: data-prepper-plugins/elasticsearch/README.md (hunk 1 above)
Comment (yadavcbala) [reply data-prepper-plugins/elasticsearch/README.md]: If the guidance you got is from Legal - I think you are right, The one i got is from a forwarded email and it was long back. There were couple of approved names given in that but the gist was to prefer Amazon Elasticsearch Service. I took it as use Amazon Elasticsearch Service :-)

Comment (wrijeff) [reply data-prepper-plugins/elasticsearch/README.md]: Cool, thanks for the info. Using the full name is never wrong so we're good in this PR 👍

**Pull request merged**